src/umode.rs

```diff
@@ -310,23 +313,19 @@ impl UmodeTask {
                 .map_err(Error::HypMap)?;
         }
         let ret = task.run();
-        if let Err(e) = &ret {
-            match e {
-                ExecError::UnexpectedTrap | ExecError::Panic => {
-                    // Reset on unrecoverable u-mode error.
-                    // 1. restore memory to original state
-                    PerCpu::this_cpu().page_table().restore_umode();
-                    // 2. setup umode again.
-                    // Panic if we can't restore umode: the hypervisor is in an unrecoverable state.
-                    task.reset().expect("Failed to recover U-mode.");
-                }
-                ExecError::Umode(_) => {}
-            }
+        // Errors encountered while executing the operation are unrecoverable: reset U-mode.
+        if ret.is_err() {
+            // 1. restore memory to original state
+            PerCpu::this_cpu().page_table().restore_umode();
```

Comment:
nit: `PerCpu::this_cpu()` gets called a bunch in this function -- pull it into a local?

Reply:
Good point, done.